### DISCUSSION TITLE
ESLint --fix adds semi-colon to json files

### DISCUSSION THREAD

**Discussion by joaocasarin** [eslint/eslint#15801] [General]
### Environment

Node version: 16.14.0
npm version: 8.7.0
Local ESLint version: 8.13.0
Global ESLint version: not using global
Operating System: Windows 10 Home Single Language

Having the rule `semi: ['error', 'always']` makes eslint add a semicolon to the end of .json files even if it is completely wrong and there is no semicolon for json files.

Error: 
```sh
PS C:\Users\joaov\Desktop\lambda-ts> yarn lint:fix
yarn run v1.22.17
warning ..\..\package.json: No license field
$ eslint --fix ./**/*.{js,ts,json}

C:\Users\joaov\Desktop\lambda-ts\package.json
  55:1  error  Parsing error: Unexpected token

C:\Users\joaov\Desktop\lambda-ts\tsconfig.json
  33:1  error  Parsing error: Unexpected token

✖ 2 problems (2 errors, 0 warnings)

error Command failed with exit code 1.
```

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```js
// .eslintrc.js
module.exports = {
    env: {
        browser: true,
        es2021: true,
        jest: true
    },
    extends: ['airbnb-base', 'prettier'],
    parser: '@typescript-eslint/parser',
    settings: {
        'import/parsers': {
            '@typescript-eslint/parser': ['.ts', '.tsx']
        },
        'import/resolver': {
            typescript: {
                alwaysTryTypes: true,
                project: './tsconfig.json'
            }
        }
    },
    parserOptions: {
        ecmaVersion: 12,
        sourceType: 'module'
    },
    plugins: ['@typescript-eslint', 'import', 'prettier'],
    ignorePatterns: ['node_modules/', 'lib/', 'coverage/'],
    rules: {
        'import/extensions': ['error', 'never'],
        'import/prefer-default-export': 'off',
        'import/first': 'off',
        'linebreak-style': ['error', 'unix'],
        indent: ['error', 4, { SwitchCase: 1, ignoredNodes: ['PropertyDefinition'] }],
        'comma-dangle': ['error', 'never'],
        'class-methods-use-this': 'off',
        semi: ['error', 'always'],
        'import/no-extraneous-dependencies': [
            'error',
            {
                devDependencies: [
                    '**/*.test.ts',
                    '**/*.test.js',
                    'serverless/**/*.ts',
                    'jest.config.js',
                    'serverless.ts',
                    '.eslintrc.js',
                    'prettier.config.js'
                ]
            }
        ]
    }
};
```
</details>

<details>
<summary>Package.json</summary>

```json
{
    "name": "lambda-ts",
    "version": "1.0.0",
    "description": "Serverless aws-nodejs-typescript template",
    "main": "serverless.ts",
    "author": "João Casarin",
    "license": "MIT",
    "scripts": {
        "test": "jest",
        "lint": "eslint --ext .js,.ts,.json",
        "lint:fix": "eslint --fix ./**/*.{js,ts,json}",
        "format": "prettier --write ./**/*.{ts,js,json}",
        "prettier:check": "prettier --check ./**/*.{ts,js,json}"
    },
    "engines": {
        "node": ">=14.15.0"
    },
    "dependencies": {
        "aws-lambda": "^1.0.7",
        "bcrypt": "^5.0.1",
        "dotenv": "^16.0.0",
        "pg": "^8.7.3",
        "reflect-metadata": "^0.1.13",
        "typeorm": "^0.3.6",
        "uuid": "^8.3.2",
        "yup": "^0.32.11"
    },
    "devDependencies": {
        "@serverless/typescript": "^3.8.0",
        "@types/aws-lambda": "^8.10.93",
        "@types/bcrypt": "^5.0.0",
        "@types/jest": "^27.4.1",
        "@types/node": "^17.0.25",
        "@types/pg": "^8.6.5",
        "@types/uuid": "^8.3.4",
        "@typescript-eslint/eslint-plugin": "^5.20.0",
        "@typescript-eslint/parser": "^5.20.0",
        "esbuild": "^0.14.36",
        "esbuild-jest": "^0.5.0",
        "eslint": "^8.13.0",
        "eslint-config-airbnb-base": "^15.0.0",
        "eslint-config-prettier": "^8.5.0",
        "eslint-import-resolver-typescript": "^2.7.1",
        "eslint-plugin-import": "^2.26.0",
        "eslint-plugin-prettier": "^4.0.0",
        "jest": "^27.5.1",
        "prettier": "^2.6.2",
        "serverless": "^3.14.0",
        "serverless-esbuild": "^1.26.2",
        "serverless-offline": "^8.7.0",
        "ts-jest": "^27.1.4",
        "ts-node": "^10.7.0",
        "typescript": "^4.6.3"
    }
}
```
</details>

<details>
<summary>Execution</summary>

Just use both codes above and run `yarn lint:fix`, then you shall get the same error as I got.
</details>

### What did you expect to happen?

Should not add the semicolon to json files.

### What actually happened?

Adds a semicolon at the end of the json file.

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by aladdin-add** [comment]
hi, eslint won't lint/fix json files, so it seems not an issue in eslint. see  the [demo](https://eslint.org/demo#eyJ0ZXh0Ijoie1wiZm9vXCI6IDF9Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLWxvc3Mtb2YtcHJlY2lzaW9uIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNldHRlci1yZXR1cm4iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW5zYWZlLW9wdGlvbmFsLWNoYWluaW5nIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjIsImNvbW1hLXN0eWxlIjoyfSwiZW52Ijp7ImJyb3dzZXIiOnRydWUsImVzMjAyMSI6dHJ1ZX19fQ==)


i'm guessing it's related to eslint intergations, e.g. ts-eslint.

**Reply by joaocasarin** [reply]
There is no ts-eslint package. Although I'm using the `eslint-import-resolver-typescript`, this "issue" started before I started using that resolver. Actually it only happens when I try to use the eslint `fix`, but when I just lint the code, eslint doesn't expect a semicolon in json files (it still tries to lint those files).

**Reply by aladdin-add** [reply]
maybe prettier? can you disable it to see if it still happens?

**Reply by joaocasarin** [reply]
My prettier is set to format on every file save and it doesn't add the semicolon to json files. I have just tried disabling prettier and running `eslint --fix` again and got the same behaviour: eslint adding the semicolon to the file.

As you can see, when I just check if eslint would complain about something, nothing appears as an error, but when I try to fix with eslint, this is the output:

![image](https://user-images.githubusercontent.com/48847394/165218409-371ab3f7-e544-4958-b986-14e2b53de2ca.png)

**Reply by aladdin-add** [reply]
emmm...that's weird! have you removed all the prettier in your eslint config?

Good to have a repo so we can look into.

**Reply by aladdin-add** [reply]
no worries, I was able to repro the issue: https://github.com/aladdin-add/eslint-issue-15801

it seems an issue in ts-eslint(it won't happen using the default parser). can you file the issue here: https://github.com/typescript-eslint/typescript-eslint?


just a reminder: you should not pass json files to eslint (as ts-eslint parser was only mean to parse ts files).

**Reply by joaocasarin** [reply]
i'm sorry taking too long for that, i had a few appointments...

yeah, that repo reproduces the behaviour exactly.
Sure, i will file it to ts-eslint repo.

Well, by 'pass json files to eslint' you mean in the --fix command? because I like that eslint complains about weird identations, unnecessary quotes and other stuffs..

**Reply by aladdin-add** [reply]
eslint is to lint JS, but JSON is not a valid JS. so you will just get a parsing error like "unexpected char....". I would suggest 2 choices:

1. just use prettier to format your json files.
2. use eslint with a custom plugin like [eslint-plugin-json](https://www.npmjs.com/package/eslint-plugin-json) to lint JSON.

**Reply by toinetoine** [reply]
@aladdin-add I don't think [eslint-plugin-json](https://www.npmjs.com/package/eslint-plugin-json) supports `--fix`: https://github.com/azeemba/eslint-plugin-json/issues/8#issuecomment-604294953

**Reply by Rec0iL99** [reply]
@toinetoine you can try using [eslint-plugin-json-format](https://github.com/kuceb/eslint-plugin-json-format). But it's quite old and not updated frequently.

**Comment by jonathanbell** [comment]
In my case, I had `json` and `jsonc` inside `"eslint.validate": {...}` in VS Code settings.

**Reply by chambliss** [reply]
Thank you, this JSON issue was driving me nuts and this was totally the problem. 😂

**Reply by gokh4nozturk** [reply]
After deleting `jsonc` inside from `"eslint.validate": {...}`, it's figured out.

**Reply by ackvf** [reply]
I wonder why I put it there in the first place! 🤦‍♂️
